src/netaddress.h

```diff
@@ -272,6 +341,48 @@ class CNetAddr
 
             UnserializeV1Array(serialized);
         }
+
+        /**
+         * Unserialize from a ADDRv2 / BIP155 format.
+         */
+        template <typename Stream>
+        void UnserializeV2Stream(Stream& s)
+        {
+            uint8_t bip155_net;
+            s >> bip155_net;
+
+            size_t address_size;
+            s >> COMPACTSIZE(address_size);
+
+            if (address_size > MAX_ADDRV2_SIZE) {
+                throw std::ios_base::failure(strprintf(
+                    "Address too long: %u > %u", address_size, MAX_ADDRV2_SIZE));
+            }
+
+            if (RecognizeBIP155Network(bip155_net, address_size)) {
+                m_addr.resize(address_size);
+                s >> MakeSpan(m_addr);
+
+                // Recognize NET_INTERNAL embedded in IPv6.
+                if (m_net == NET_IPV6 && HasPrefix(m_addr, INTERNAL_IN_IPV6_PREFIX)) {
```

Comment:
Why do this for NET_INTERNAL, but not for IPv4 or TORv2?

Reply:
Because IPv4 and TORv2 have their own BIP155 network ids and should not be disguised as IPv6.

`NET_INTERNAL` is not gossiped, but addrman may keep such addresses as "source" of other addresses and thus end up (un)serializing them when saving its database to disk in a V2 stream (not in this PR).

In this PR we may end up having some TORv3 addresses in addrman and try to serialize them in V1 stream. They will go as 16 zeroes.

Reply:
Following an [IRC discussion](http://www.erisian.com.au/bitcoin-core-dev/log-2020-09-11.html#l-406) - I changed it to recognize and ignore (unserialize as invalid) embedded IPv4 and TORv2 in IPv6 (when in addrv2 message).